[1] pipeline - fix it

**Describe the bug**
The pipeline is failing constantly with various errors.
Some errors:
- Run kubespray
```
TASK [kubernetes/kubeadm : Check if kubeadm CA cert is accessible] *************
fatal: [pipeline-wc-master-0 -> pipeline-wc-master-0]: FAILED! => {"msg": "Timeout (12s) waiting for privilege escalation prompt: "}
```

- Install apps
```
Run ./apps/pipeline/apply-sc.bash
  ./apps/pipeline/apply-sc.bash
  shell: bash --noprofile --norc -e -o pipefail {0}
  env:
    EXOSCALE_API_KEY: ***
    EXOSCALE_API_SECRET: ***
    CK8S_CONFIG_PATH: ./apps/pipeline/config/exoscale
    CK8S_AUTO_APPROVE: true
panic: runtime error: index out of range [5] with length 5
```

**Expected behaviour**
Green pipeline.

**Definition of done**
- fix the above errors and ensure that the end-to-end pipeline is green, for multiple run
- fix the secrets issues https://github.com/elastisys/compliantkubernetes-apps/issues/184

--- a/.github/workflows/end-to-end.yml
+++ b/.github/workflows/end-to-end.yml
@@ -91,7 +91,7 @@ jobs:
         args: route53 change-resource-record-sets --hosted-zone-id Z2STJRQSJO5PZ0 --change-batch file://apps/pipeline/config/exoscale/dns-${{ matrix.cluster }}.json
 
     - name: Run kubespray
-      uses: docker://quay.io/kubespray/kubespray:v2.18.0
+      uses: docker://quay.io/kubespray/kubespray:v2.19.0
       env:
         ANSIBLE_CONFIG: ./compliantkubernetes-kubespray/kubespray/ansible.cfg
       with:

--- a/.github/workflows/pipeline.yml
+++ b/.github/workflows/pipeline.yml
@@ -18,21 +18,17 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        cloud: ["safespring", "exoscale", "citycloud", "aws"]
+        cloud: ["safespring", "exoscale", "citycloud", "elastx", "upcloud"]
         flavor: ["prod", "dev"]
     steps:
     - name: Checkout compliantkubernetes-apps
       uses: actions/checkout@v2
       with:
         fetch-depth: 0 # Fetches all tags
-    - name: Setup pgp key for sops
-      uses: ./pipeline
-      with:
-        args: ./pipeline/sops-pgp-setup.bash
     - name: Init and configure config for compliantkubernetes-apps
       uses: ./pipeline
       with:
-        args: ./pipeline/init-unit-tests.bash
+        args: bash -c "./pipeline/sops-pgp-setup.bash && ./pipeline/init-unit-tests.bash"
       env:
         CK8S_CLOUD_PROVIDER: ${{ matrix.cloud }}
         CI_CK8S_FLAVOR: ${{ matrix.flavor }}
@@ -40,7 +36,7 @@ jobs:
     - name: Run linters
       uses: ./pipeline
       with:
-        args: ./pipeline/lint.bash
+        args: bash -c "./pipeline/sops-pgp-setup.bash && ./pipeline/lint.bash"
 
   opa:
     runs-on: ubuntu-latest
@@ -54,6 +50,8 @@ jobs:
 
   bats:
     runs-on: ubuntu-latest
+    env:
+      CK8S_PGP_FP: 529D964DE0BBD900C4A395DA09986C297F8B7757
     steps:
       - name: Checkout compliantkubernetes-apps
         uses: actions/checkout@v2

--- a/pipeline/Dockerfile
+++ b/pipeline/Dockerfile
@@ -1,4 +1,4 @@
-FROM ubuntu:18.04
+FROM ubuntu:20.04
 
 RUN  apt-get update && \
      apt-get install -y software-properties-common && \
@@ -12,7 +12,7 @@ RUN  apt-get update && \
      rm -rf /var/lib/apt/lists/*
 
 # Kubectl
-ENV KUBECTL_VERSION "v1.20.7"
+ENV KUBECTL_VERSION "v1.22.6"
 RUN wget "https://storage.googleapis.com/kubernetes-release/release/${KUBECTL_VERSION}/bin/linux/amd64/kubectl" && \
     chmod +x ./kubectl && \
     mv ./kubectl /usr/local/bin/kubectl
@@ -25,7 +25,7 @@ RUN wget https://github.com/mozilla/sops/releases/download/v${SOPS_VERSION}/sops
     chmod +x /usr/local/bin/sops
 
 # Helm
-ENV HELM_VERSION "v3.5.2"
+ENV HELM_VERSION "v3.8.1"
 RUN wget "https://get.helm.sh/helm-${HELM_VERSION}-linux-amd64.tar.gz" && \
     tar -zxvf "helm-${HELM_VERSION}-linux-amd64.tar.gz" && \
     mv linux-amd64/helm /usr/local/bin/helm && \
@@ -35,10 +35,10 @@ RUN wget "https://get.helm.sh/helm-${HELM_VERSION}-linux-amd64.tar.gz" && \
 # Should come with v3.3.0, see https://github.com/helm/helm/pull/7983
 ENV XDG_DATA_HOME=/root/.config
 RUN helm plugin install https://github.com/databus23/helm-diff --version v3.1.2
-RUN helm plugin install https://github.com/futuresimple/helm-secrets --version v2.0.2
+RUN helm plugin install https://github.com/jkroepke/helm-secrets --version v3.12
 
 # Helmfile
-ENV HELMFILE_VERSION "v0.138.4"
+ENV HELMFILE_VERSION "v0.144.0"
 RUN wget "https://github.com/roboll/helmfile/releases/download/${HELMFILE_VERSION}/helmfile_linux_amd64" && \
     chmod +x helmfile_linux_amd64 && \
     mv helmfile_linux_amd64 /usr/local/bin/helmfile

--- a/pipeline/config/exoscale/common-config.yaml
+++ b/pipeline/config/exoscale/common-config.yaml
@@ -5,11 +5,9 @@ global:
   baseDomain: pipeline-exoscale.elastisys.se
   opsDomain: ops.pipeline-exoscale.elastisys.se
 objectStorage:
-  type: s3
   s3:
     region: ch-gva-2
     regionEndpoint: https://sos-ch-gva-2.exo.io
-    forcePathStyle: true
 ingressNginx:
   controller:
     service:
@@ -29,6 +27,3 @@ clusterAdmin:
 monitoring:
   rook:
     enabled: true
-prometheus:
-  s3BucketPercentLimit: 80
-  s3BucketSizeQuotaGB: 1000

--- a/pipeline/config/exoscale/pipeline-sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster-default.yaml
+++ b/pipeline/config/exoscale/pipeline-sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster-default.yaml
@@ -1,1 +1,2 @@
-etcd_kubeadm_enabled: true
+etcd_deployment_type: kubeadm
+kubeconfig_localhost_ansible_host: true

--- a/pipeline/config/exoscale/pipeline-sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml
+++ b/pipeline/config/exoscale/pipeline-sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml
@@ -2,6 +2,7 @@ kube_oidc_auth: true
 kube_oidc_url: "https://dex.pipeline-exoscale.elastisys.se"
 kube_oidc_client_id: "kubelogin"
 kube_oidc_username_claim: "email"
+kube_oidc_groups_claim: "groups"
 
 kube_apiserver_enable_admission_plugins:
   - "PodSecurityPolicy"
@@ -92,3 +93,10 @@ audit_policy_custom_rules: |-
       - "RequestReceived"
 podsecuritypolicy_enabled: true
 kubeconfig_localhost: true
+calico_felix_prometheusmetricsenabled: true
+
+resolvconf_mode: host_resolvconf
+container_manager: containerd
+kubelet_config_extra_args:
+  imageGCHighThresholdPercent: 75
+  imageGCLowThresholdPercent: 70

--- a/pipeline/config/exoscale/pipeline-wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster-default.yaml
+++ b/pipeline/config/exoscale/pipeline-wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster-default.yaml
@@ -1,1 +1,2 @@
-etcd_kubeadm_enabled: true
+etcd_deployment_type: kubeadm
+kubeconfig_localhost_ansible_host: true

--- a/pipeline/config/exoscale/pipeline-wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml
+++ b/pipeline/config/exoscale/pipeline-wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml
@@ -2,6 +2,7 @@ kube_oidc_auth: true
 kube_oidc_url: "https://dex.pipeline-exoscale.elastisys.se"
 kube_oidc_client_id: "kubelogin"
 kube_oidc_username_claim: "email"
+kube_oidc_groups_claim: "groups"
 
 kube_apiserver_enable_admission_plugins:
   - "PodSecurityPolicy"
@@ -92,3 +93,10 @@ audit_policy_custom_rules: |-
       - "RequestReceived"
 podsecuritypolicy_enabled: true
 kubeconfig_localhost: true
+calico_felix_prometheusmetricsenabled: true
+
+resolvconf_mode: host_resolvconf
+container_manager: containerd
+kubelet_config_extra_args:
+  imageGCHighThresholdPercent: 75
+  imageGCLowThresholdPercent: 70

--- a/pipeline/config/exoscale/sc-config.yaml
+++ b/pipeline/config/exoscale/sc-config.yaml
@@ -13,7 +13,6 @@ harbor:
     adminGroupName: not-used
   backup:
     enabled: true
-    retentionDays: 7
 opensearch:
   masterNode:
     storageSize: 1Gi
@@ -25,26 +24,11 @@ opensearch:
       limits:
         memory: 1024Mi
         cpu: 1
-  dataNode:
-    dedicatedPods: true
-    storageSize: 18Gi
-  clientNode:
-    dedicatedPods: true
-  curator:
-    retention:
-      authLogAgeDays: 7
   snapshot:
     enabled: true
   extraRoleMappings: []
 fluentd:
   enabled: true
-alerts:
-  opsGenieHeartbeat:
-    name: not-used
-  slack:
-    channel: not-used
 s3Exporter:
   interval: 120s
   scrapeTimeout: 30s
-prometheus:
-  predictLinearLimit: 66

--- a/pipeline/config/exoscale/wc-config.yaml
+++ b/pipeline/config/exoscale/wc-config.yaml
@@ -5,9 +5,6 @@ opa:
   imageRegistry:
     URL:
       - harbor.pipeline-exoscale.elastisys.se
-issuers:
-  letsencrypt:
-    enabled: true
 user:
   namespaces:
     - demo1
@@ -16,5 +13,3 @@ user:
   adminUsers:
     - user-admin@example.com
   adminGroups: []
-prometheus:
-  predictLinearLimit: 66

--- a/pipeline/init-unit-tests.bash
+++ b/pipeline/init-unit-tests.bash
@@ -4,7 +4,7 @@ set -eu -o pipefail
 
 here="$(dirname "$(readlink -f "$0")")"
 ck8s="${here}/../bin/ck8s"
-# shellcheck source=bin/common.bash
+# shellcheck source=pipeline/common.bash
 source "${here}/common.bash"
 
 export CK8S_FLAVOR="${CI_CK8S_FLAVOR:-dev}"

--- a/pipeline/init.bats
+++ b/pipeline/init.bats
@@ -13,7 +13,6 @@ setup() {
     export CK8S_CONFIG_PATH="${ck8s_config_path_tmp}"
     export CK8S_ENVIRONMENT_NAME=test
     export CK8S_FLAVOR=dev
-    export CK8S_PGP_FP=529D964DE0BBD900C4A395DA09986C297F8B7757
     export CK8S_CLOUD_PROVIDER=aws
 }
 

--- a/pipeline/lint.bash
+++ b/pipeline/lint.bash
@@ -2,10 +2,9 @@
 
 set -eu -o pipefail
 
-here="$(dirname "$(readlink -f "$0")")"
+here="$(dirname "$(readlink -f "${0}")")"
 
 #We should only lint our "own" charts, not the upstream ones
-# TODO: enable linting of kubeapi-metrics, opensearch-secrets and opensearch-configurer after the PGP issues in the pipeline have been solved (https://github.com/elastisys/compliantkubernetes-apps/issues/184)
 
 charts_ignore_list=(
   "app!=sc-logs-retention"
@@ -15,7 +14,6 @@ charts_ignore_list=(
   "app!=kube-prometheus-stack"
   "app!=velero"
   "app!=metrics-server"
-  "app!=dex"
   "app!=wc-scraper"
   "app!=prometheus-auth"
   "app!=wc-reader"
@@ -26,17 +24,11 @@ charts_ignore_list=(
   "app!=thanos"
   "app!=fluentd"
   "app!=fluentd-aggregator"
-  "app!=opensearch-master"
   "app!=opensearch-data"
   "app!=opensearch-client"
-  "app!=opensearch-dashboards"
-  "app!=opensearch-secrets"
-  "app!=opensearch-securityadmin"
-  "app!=opensearch-configurer"
   "app!=falco"
   "app!=falco-exporter"
-  "app!=user-alertmanager"
-  "app!=kubeapi-metrics")
+  "app!=user-alertmanager")
 
 helmfile -e service_cluster -f "${here}/../helmfile/" -l "$(IFS=',' ; echo "${charts_ignore_list[*]}")" lint
 helmfile -e workload_cluster -f "${here}/../helmfile/" -l "$(IFS=',' ; echo "${charts_ignore_list[*]}")" lint

--- a/pipeline/test/services/service-cluster/testPrometheusTargets.sh
+++ b/pipeline/test/services/service-cluster/testPrometheusTargets.sh
@@ -10,6 +10,8 @@ source "$INNER_SCRIPTS_PATH/../prometheus-common.sh"
 totalNodes=$(kubectl get nodes --no-headers | wc -l)
 masterNodes=$(kubectl get nodes -l node-role.kubernetes.io/master --no-headers | wc -l)
 
+totalPrometheus=$(yq r -e "${CONFIG_FILE}" 'prometheus.replicas')
+
 enable_thanos=$(yq r -e "${CONFIG_FILE}" 'thanos.enabled')
 enable_thanos_receiver=$(yq r -e "${CONFIG_FILE}" 'thanos.receiver.enabled')
 enable_thanos_query=$(yq r -e "${CONFIG_FILE}" 'thanos.query.enabled')
@@ -31,7 +33,7 @@ scTargets=(
     "serviceMonitor/monitoring/prometheus-blackbox-exporter-dex/0 1"
     "serviceMonitor/monitoring/prometheus-blackbox-exporter-grafana/0 1"
     "serviceMonitor/monitoring/prometheus-blackbox-exporter-opensearch-dashboards/0 1"
-    "serviceMonitor/monitoring/kube-prometheus-stack-alertmanager/0 1"
+    "serviceMonitor/monitoring/kube-prometheus-stack-alertmanager/0 2"
     "serviceMonitor/monitoring/kube-prometheus-stack-apiserver/0 ${masterNodes}"
     "serviceMonitor/monitoring/kube-prometheus-stack-coredns/0 2"
     "serviceMonitor/monitoring/kube-prometheus-stack-grafana/0 1"
@@ -40,7 +42,7 @@ scTargets=(
     "serviceMonitor/monitoring/kube-prometheus-stack-kubelet/1 ${totalNodes}"
     "serviceMonitor/monitoring/kube-prometheus-stack-node-exporter/0 ${totalNodes}"
     "serviceMonitor/monitoring/kube-prometheus-stack-operator/0 1"
-    "serviceMonitor/monitoring/kube-prometheus-stack-prometheus/0 1"
+    "serviceMonitor/monitoring/kube-prometheus-stack-prometheus/0 ${totalPrometheus}"
 )
 if [[ "${enable_thanos}" == "true" ]] && [[ "${enable_thanos_service_monitor}" == "true" ]] && [[ "${enable_thanos_receiver}" == "true" ]]; then
     scTargets+=(

